File: backend/controllers/checkoutController.js
import Stripe from 'stripe';

const stripe = new Stripe(process.env.STRIPE_SECRET_KEY || '', {
  apiVersion: '2023-10-16',
});

export const createCheckoutSession = async (req, res) => {
  try {
    const { lineItems, amount } = req.body;
    if ((!lineItems || lineItems.length === 0) && !amount) {
      return res.status(400).json({ message: 'Line items or amount is required' });
    }

    const session = await stripe.checkout.sessions.create({
      payment_method_types: ['card'],
      mode: 'payment',
      line_items:
        lineItems && lineItems.length > 0
          ? lineItems
          : [
              {
                price_data: {
                  currency: 'usd',
                  product_data: { name: 'E-commerce order' },
                  unit_amount: amount,
                },
                quantity: 1,
              },
            ],
      success_url: process.env.CHECKOUT_SUCCESS_URL || 'http://localhost:5173/checkout?status=success',
      cancel_url: process.env.CHECKOUT_CANCEL_URL || 'http://localhost:5173/checkout?status=cancelled',
    });

    return res.json({ url: session.url, sessionId: session.id });
  } catch (error) {
    console.error('Stripe checkout error:', error);
    return res.status(500).json({ message: error.message || 'Stripe initialization failed' });
  }
};

File: backend/app/controllers/api/checkout_controller.rb
# frozen_string_literal: true

module Api
  class CheckoutController < BaseController
    def create_session
      line_items = params[:line_items] || params[:items]
      return render json: { error: 'line_items are required' }, status: :unprocessable_entity unless line_items.present?

      normalized_items = normalize_line_items(line_items)
      return render json: { error: 'line_items are required' }, status: :unprocessable_entity if normalized_items.empty?
      success_url = params.fetch(:success_url, 'http://localhost:5173/success')
      cancel_url = params.fetch(:cancel_url, 'http://localhost:5173/cancel')

      stripe_session = Stripe::Checkout::Session.create(
        payment_method_types: ['card'],
        mode: 'payment',
        customer_email: current_user.email,
        line_items: normalized_items.map { |item| build_line_item(item) },
        success_url: success_url,
        cancel_url: cancel_url
      )

      render json: { id: stripe_session.id }
    rescue Stripe::StripeError => e
      render json: { error: e.message }, status: :unprocessable_entity
    end

    private

    def build_line_item(item)
      product = Product.find(item[:product_id])
      {
        price_data: {
          currency: 'usd',
          product_data: { name: product.title },
          unit_amount: (product.price * 100).to_i
        },
        quantity: item[:quantity]
      }
    end

    def normalize_line_items(line_items)
      line_items.map do |item|
        product_identifier = item[:product_id] || item[:product] || item['product_id'] || item['product']
        product_id = product_identifier.to_s
        next unless product_id.match?(/\A\d+\z/)
        product_id = product_id.to_i
        quantity = (item[:quantity] || item['quantity'] || 1).to_i

        {
          product_id: product_id,
          quantity: quantity,
          price: item[:price] || item['price']
        }.compact
      end.compact
    end
  end
end
